### PR TITLE
fix: keycode utilities

### PR DESCRIPTION
**Pull request by mathisscott**
• removed keycode enums
• added keycode service to manage registry of keycodes
• allows users to add their own keycodes
• added onKey, onAnyKey, and onKeyCombo utility functions
• added isStringAndNotNilOrEmpty utility function
• added Maybe monadic data structure to help with defensive coding
• updated button and modal to use new keycode utilities
• adding deepClone utility fn
• fixed issue with Maybe mappings trying to turn bad function returns Just
• added isMap utility function to identity.ts
• improved deepClone() utility function to handle maps, arrays, and objects
• cleaning up some warnings and issues with clr/react

Signed-off-by: Scott Mathis <smathis@vmware.com>

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [X] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] clarity.design website / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Previously we were duplicating keycode event checks across multiple files. The logic was implemented in an inconsistent fashion (see clr/angular) and what's worse is we were checking for the wrong things when it came to spacebars in some situations.

This change consolidates that logic into one place. DRYs it all up and creates a potential integration point with clr/angular.

Issue Number: #4487 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/internal/services/keycodes.service.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { Maybe } from '../data-structures/maybe.js';
+import { deepClone } from '../utils/identity.js';
+
+export enum KeyCodeTypes {
```
Comment (bbogdanov) [inline packages/core/src/internal/services/keycodes.service.ts:10]: What do you think about using `const enum` instead of `enum`?

Comment (mathisscott) [reply packages/core/src/internal/services/keycodes.service.ts]: Digging through it, it appears we don't need a lookup table so using `const` would be fine. In addition, I found that this enum is only used inside this file so we don't need to export it.

File: packages/core/src/internal/services/keycodes.service.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+export function getKeycodeFromRegistry(
+  codeToLookup: string,
+  whichCodeType: 'code' | 'ie-code' = KeyCodeTypes.DEFAULT,
+  registry = keyCodeRegistry
+): string {
+  return Maybe.of(registry.get(codeToLookup))
+    .map(mbMapVal => {
+      return (mbMapVal.get && mbMapVal.get(whichCodeType)) || null;
+    })
+    .valueOr('').unsafeValue;
```
Comment (coryrylan) [inline packages/core/src/internal/services/keycodes.service.ts:138]: nit: could this be simplified if the maybe works with undefined?

```typescript
 return Maybe.of(registry.get(codeToLookup))
    .map(mbMapVal => mbMapVal?.get(whichCodeType))
    .valueOr('').unsafeValue;
```

Comment (mathisscott) [reply packages/core/src/internal/services/keycodes.service.ts:138]: Yep. Good suggestion!

File: packages/core/src/internal/utils/keycodes.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  if (keyArray.length === 0) {
+    return true;
+  }
+
+  const curriedChecker = curryN(2, keyWasEvented);
+  const myCheckerFn = curriedChecker(evt);
+
+  return keyArray.reduce((acc: boolean, currentKey: string) => {
+    return acc !== false && myCheckerFn(currentKey);
+  }, true);
```
[the remark this answers is not in the document]
Comment (mathisscott) [reply packages/core/src/internal/utils/keycodes.ts:148]: Yeah, I think that will work. I will give it a try. Thanks!

**Pull request merged**